src/parse.js

```diff
@@ -319,13 +319,15 @@ function parseReferences(cell, input, globals = defaultGlobals) {
   return cell;
 }
 
-// Find references.
+// Find features: file attachments, secrets, database clients.
 // Check for illegal references to arguments.
 // Check for illegal assignments to global references.
-function parseFileAttachments(cell, input) {
+function parseFeatures(cell, input) {
   if (cell.body && cell.body.type !== "ImportDeclaration") {
     try {
-      cell.fileAttachments = findFileAttachments(cell);
+      cell.fileAttachments = findFeatures(cell, "FileAttachment", "FileAttachment() requires a single literal string argument.");
+      cell.databaseClients = findFeatures(cell, "DatabaseClient", "DatabaseClient() requires a single literal string argument.");
+      cell.secrets = findFeatures(cell, "Secret", "Secret() requires a single literal string argument.");
     } catch (error) {
```

Comment:
This is now performing three full walks of the AST: one for each feature we’re trying to detect. Especially considering that they’re all called statically with the same requirements, requiring a single literal string argument, do you think that we should be parsing features in a single pass instead? Or should we not be concerned about parse performance here for large cells?

Reply:
I added a benchmark. On a small cell, it makes no discernible difference (within the noise of the benchmark). On a cell that approaches the limit (60+KB), it appears to make a difference of a few milliseconds: ~33ms for a single pass, ~39ms for three passes. I think it’s probably not a high priority to optimize here, but I’ll think a little more about what a single pass might look like.